src/apple/device/mod.rs

```diff
@@ -0,0 +1,202 @@
+        self
+    }
+
+    pub fn target(&self) -> &'a Target<'a> {
+        self.target
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn model(&self) -> &str {
+        &self.model
+    }
+
+    pub fn run(
+        &self,
+        config: &Config,
+        env: &Env,
+        noise_level: opts::NoiseLevel,
+        non_interactive: bool,
+        profile: opts::Profile,
+    ) -> Result<duct::Handle, RunError> {
+        // TODO: These steps are run unconditionally, which is slooooooow
+        println!("Building app...");
+        self.target
+            .build(config, env, noise_level, profile)
+            .map_err(RunError::BuildFailed)?;
+        println!("Archiving app...");
+        self.target
+            .archive(config, env, noise_level, profile, None)
+            .map_err(RunError::ArchiveFailed)?;
+
+        match self.kind {
+            DeviceKind::Simulator => simctl::run(config, env, non_interactive, &self.id)
+                .map_err(|e| RunError::DeployFailed(e.to_string())),
+            DeviceKind::IosDeployDevice | DeviceKind::DeviceCtlDevice => {
+                println!("Exporting app...");
+                self.target
+                    .export(config, env, noise_level)
+                    .map_err(RunError::ExportFailed)?;
+                println!("Extracting IPA...");
+
+                let ipa_path = config
+                    .ipa_path()
+                    .map_err(|(old, new)| RunError::IpaMissing { old, new })?;
+                let export_dir = config.export_dir();
+                let cmd = duct::cmd::<&str, [String; 0]>("unzip", [])
+                    .vars(env.explicit_env())
+                    .before_spawn(move |cmd| {
+                        if noise_level.pedantic() {
+                            cmd.arg("-q");
+                        }
+                        cmd.arg("-o").arg(&ipa_path).arg("-d").arg(&export_dir);
+                        Ok(())
+                    });
+
+                cmd.run().map_err(RunError::UnzipFailed)?;
+
+                if self.kind == DeviceKind::IosDeployDevice {
+                    ios_deploy::run_and_debug(config, env, non_interactive, &self.id)
+                        .map_err(|e| RunError::DeployFailed(e.to_string()))
+                } else {
+                    devicectl::run(config, env, non_interactive, &self.id, self.paired)
+                        .map_err(|e| RunError::DeployFailed(e.to_string()))
+                }
+            }
+        }
+    }
+}
+
+pub fn list_devices<'a>(env: &Env) -> Result<BTreeSet<Device<'a>>, String> {
+    let mut devices = BTreeSet::default();
+    let mut error = None;
+
+    // devicectl
+    match devicectl::device_list(env) {
+        Ok(d) => {
+            devices.extend(d);
+        }
+        Err(e) => {
+            error = Some(e);
+        }
+    }
+
+    // if we could not find a device with devicectl, let's use ios-deploy
+    if devices.is_empty() {
```

Comment:
Are we able to choose `devicectl` or `ios-deploy` by xcode version or maybe `xcrun -f devicectl` and read the exit code?
So that if there's no device found by `devicectl`, it will show no device instead of installing and asking `ios-deploy`

Reply:
devicectl can only detect iOS 17 devices AFAIK and from my tests, so that's why we need to check with ios-deploy regardless.

Reply:
oh.. okay 🫠

Do we need to list devices from `devicectl` and `ios-deploy` at the same time if they have devices with iOS17 and below connected?

Reply:
my idea was to avoid installing ios-deploy at all in this case. what are the odds you'd have two devices connected at the same time?